### ISSUE TITLE
[RC1] Microsoft Identity Platform 2.0 for Blazor

### ISSUE THREAD

**Issue by guardrex** [dotnet/AspNetCore.Docs#19503]
**Update blazor templates with Microsoft.Identity.Web APIs**
https://github.com/dotnet/aspnetcore/pull/24268

**Remove accepted scope check for non-Identity.Web scenarios**
https://github.com/dotnet/aspnetcore/pull/24894

**Add Microsoft.Identity.Web migration guide to 3.1->5.0 documentation**
https://github.com/dotnet/AspNetCore.Docs/issues/19810

**Update Azure Active Directory documentation**
https://github.com/dotnet/AspNetCore.Docs/issues/19809

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: c572618c-8b98-95af-6a75-1f8e413071db
* Version Independent ID: 8670b314-460d-3194-c53b-9a44f39c6b2a
* Content: [Secure ASP.NET Core Blazor WebAssembly](https://docs.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/?view=aspnetcore-3.1)
* Content Source: [aspnetcore/blazor/security/webassembly/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/security/webassembly/index.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

**Comment by guardrex** [comment]
@captainsafia ... I have a quick point of personal confusion here as I get into the updates. I decided to start with the small footprint coverage of Blazor Server docs. These will be quick to update, and I should have a PR up tomorrow morning.

According to [John's cross-link](https://github.com/dotnet/AspNetCore.Docs/issues/19810) to the AAD docs ...

**Quickstart: Add sign-in with Microsoft to an ASP.NET Core web app**
https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-v2-aspnet-core-webapp

... we're still setting "**_Under Implicit grant, select ID tokens._**" ... but I thought we were getting out of the business of implicit grant completely for 5.0 scenarios. My 5.0 Blazor Server with AAD test app won't run unless I provide it with implicit grant for `id_token`. With it, our [coverage for passing tokens](https://docs.microsoft.com/en-us/aspnet/core/blazor/security/server/additional-scenarios?view=aspnetcore-3.1#pass-tokens-to-a-blazor-server-app) also works fine. Just wondering if you could clarify what I missed about dropping implicit grant ... i.e. ... was it **only** being dropped for client-side now (i.e., SPA platform configuration with Blazor WASM scenarios)?

I also note in passing that we should cross-link that AAD topic for Blazor Server. We don't carry Blazor Server content, which is good, but we also don't link to that Quickstart. In addition to that *Web App* quickstart cross-link, I'll also add a cross-link for ...

**Quickstart: Protect an ASP.NET Core web API with Microsoft identity platform**
https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-v2-aspnet-core-web-api

**Comment by captainsafia** [comment]
>  i.e. ... was it only being dropped for client-side now (i.e., SPA platform configuration with Blazor WASM scenarios)?

Correct. It's only being dropped for SPA-scenarios.

**Comment by guardrex** [comment]
@captainsafia Hit a new one this morning ... it pertains to default scopes for a standalone Azure AD/MS accounts WASM app ...

* [empty_input_scopes_error error when using MSAL Auth (dotnet/aspnetcore #25961)](https://github.com/dotnet/aspnetcore/issues/25961)
* [Pass all scopes in token request and update IndividualLocalAuth template (dotnet/aspnetcore #26232)](https://github.com/dotnet/aspnetcore/pull/26232)

For 3.x with a standalone WASM app, it wasn't necessary to specify a default scope. If the guidance is followed for a 5.x app without taking additional action, the app throws ...

> ClientConfigurationError: empty_input_scopes_error: Scopes cannot be passed as null, undefined or empty array because they are required to obtain an access token.

So, I can give it an `openid` scope ...

```csharp
options.ProviderOptions.DefaultAccessTokenScopes.Add("openid");
```

... and the app seems happy with that. If that's the right path, what about `offline_access`? Should it also specify `offline_access` in the app as an additional scope? ...

```csharp
options.ProviderOptions.AdditionalScopesToConsent.Add("offline_access");
```

**Comment by captainsafia** [comment]
> Question 1

@guardrex With regard to the default scopes issue, yes, we need this included in the `Program.cs`. It makes sense to extend it in [this portion of the docs](https://docs.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/standalone-with-azure-active-directory?view=aspnetcore-3.1#access-token-scopes). 

> Question 2

We'll need to add the `offline_access` scope to support refresh tokens. It shouldn't matter but would recommend `DefaultAccessTokenScopes` to reduce concept count.

> Question 3

I think keeping the setup for 3.x and modifying only 5.0 might reduce confusion, especially for folks who already have 3.x apps running.

> Question 4

Behind the scenes, both default and additional scopes get propagated into the requests. The difference is largely a conceptual one, IMO. Default scopes for things like `openid`, `email`, `profile`, and `offline_access` and default API scopes for anything else and additional scopes for other values.

We use `DefaultScopes` in all our docs currently so we might want to continue with that pattern.

**Comment by guardrex** [comment]
cc: @captainsafia ... I'll need one more day depending on working out this last scenario.

@jmprieur ... May I ask for a tip? ... I'm stuck on the last section update for Blazor WASM security with Identity v2.0 and Graph API/SDK.

If you're aware of a sample that's fairly close to what I'm trying to accomplish, feel free to just post back with the link. I'm fairly good at hacking ........ er ... *programming* :smile: ....... based on samples that are close to my goals.

TL;DR ... Here's what we have ...

Blazor WASM app in a hosted Blazor solution ...

* The clientside WASM app can authenticate the user. Working fine. :+1:
* The clientside WASM app can get Azure security group memberships and based on policies permit/deny access to parts of the app based on those groups. Working fine. :+1:
* The client WASM app can call a the server API app in the server API app for weather data ... as we've been doing for quite a while in our demo apps. This is working fine now that you solved my `Authority` problem because the tenant publisher domain isn't validated (so I get the `https://`-based scope). There's content now in our docs to explain the tenant publisher domain influence on the configuration, so I think readers will be ok going forward .... anyway .... this piece is working fine. :+1:
* **_The server API app should also make a Graph API call to obtain the user's security groups so that server-side policies can permit/deny access to API endpoints beyond simple authn._** :boom: ... This is where things aren't quite working yet.

Pseudo/hacked code goes like this in the server API app's `Startup.ConfigureServices` ...

```csharp
JwtSecurityTokenHandler.DefaultMapInboundClaims = false;

services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
    .AddMicrosoftIdentityWebApi(options =>
{
    Configuration.Bind("AzureAd", options);
    options.Events = new JwtBearerEvents()
    {
        OnTokenValidated = async context =>
        {
            var graphClient = (GraphServiceClient)context.HttpContext.RequestServices
                .GetService(typeof(GraphServiceClient));
            var accessToken = context.SecurityToken as JwtSecurityToken;
            var oid = accessToken.Claims.FirstOrDefault(x => x.Type == "oid")?.Value;

            if (!string.IsNullOrEmpty(oid))
            {
                var userIdentity = (ClaimsIdentity)context.Principal.Identity;

                IUserMemberOfCollectionWithReferencesPage groupsAndAzureRoles = null;

                try
                {
                    groupsAndAzureRoles = await graphClient.Users[oid].MemberOf.Request().GetAsync();
                }
                catch (ServiceException serviceException)
                {
                    // Log the error
                }

                if (groupsAndAzureRoles != null)
                {
                    foreach (var entry in groupsAndAzureRoles)
                    {
                        userIdentity.AddClaim(new Claim("group", entry.Id));
                    }
                }
            }

            await Task.FromResult(0);
        }
    };
},
options =>
{
    Configuration.Bind("AzureAd", options);
})
    .EnableTokenAcquisitionToCallDownstreamApi(options => Configuration.Bind("AzureAd", options))
        .AddMicrosoftGraph(Configuration.GetSection("Graph"))
    .AddInMemoryTokenCaches();

services.AddControllersWithViews();
services.AddRazorPages();

services.AddAuthorization(options =>
{
    options.AddPolicy("BillingAdmin", policy => policy.RequireClaim("group", "69ff516a-b57d-4697-a429-9de4af7b5609"));
});
```

... with settings like this ...

```json
"AzureAd": {
    "Instance": "https://login.microsoftonline.com/",
    "Domain": "XXXXXXXXXXXX.onmicrosoft.com",
    "TenantId": "{TENANT ID GUID}",
    "ClientId": "{CLIENT ID GUID}",
    "CallbackPath": "/signin-oidc",
    "Audience": "https://XXXXXXXXXXXX.onmicrosoft.com/{CLIENT ID}",
    "ClientSecret": "{SECRET}"
  },
  "Graph": {
    "BaseUrl": "https://graph.microsoft.com/v1.0",
    "Scopes": "directory.read.all"
  },
  ...
}
```

The error is ........ 

> Service Exception: IDW10502: An MsalUiRequiredException was thrown due to a challenge for the user. See https://aka.ms/ms-id-web/ca_incremental-consent.

Implying that it wants user UI consent because the scopes are incorrect ...

> If, at some point, your app requests more scopes than what the admin has consented, you'll receive an MsalUiRequiredException, and you'll know that you need to have more scopes pre-approved by the tenant admin.

... but prior apps under ADAL/Identity v1.0 only seemed to require `Directory.Read.All` for this approach to work. I can hack around some more with my scopes ... requested and assigned ... if you think that's where I've gone wrong here.

Whatever the problem here :point_up:, there are two other side-questions that may or may not be relevant ...

1. Some of the Azure examples have `EnableTokenAcquisitionToCallDownstreamApi` with a parameter list that only includes an `initialScopes` string array. However when I try that, it throws a `cannot convert from 'string[]' to 'System.Action<Microsoft.Identity.Client.ConfidentialClientApplicationOptions>'` ...

   ```csharp
   string[] initialScopes = Configuration.GetValue<string>("Graph:Scopes")?.Split(' ');
   ```

   ... with `Graph:Scopes` set as `"directory.read.all"`.

1. At least one of the Azure examples has in `OnTokenValidated` ...

   ```csharp
   if (context.Principal.Claims.Any(x => (x.Type == "azp" || x.Type == "appid") && x.Value != Configuration["AzureAd:ClientId"]))
   {
       throw new UnauthorizedAccessException("Client is not authorized to access the resource.");
    }
   else
   {
       throw new UnauthorizedAccessException("Application ID claim does not exist for the client.");
   }
   ```

   ... but I don't see how that can work. It has an `else` and seems to throw no matter whether the test is `true` or `false`. btw - In my scenario, that test always seems to evaluate `false`.

**Comment by ChristianWeyer** [comment]
How much will .NET 5 and Identity Platform 2.0 influence the existing code people are writing with Blazor WASM on .NET Core 3.1, e.g. when using IdentityServer? @guardrex

**Comment by jmprieur** [comment]
@guardrex : are you looking for a Blazor Server app calling graph (it's done by the templates when writing: 
`dotnet new blazorserver --auth SingleOrg --calls-graph`

And a sample is available from https://github.com/AzureAD/microsoft-identity-web/tree/master/tests/BlazorServerCallsGraph

or a Blazorwasm application?
- not hosted: @javiercn has the code.
- hosted, it's available from the Microsoft.Identity.Web templates (`dotnet new blazorwasm2 --auth SingleOrg --calls-graph --hosted`)

**Comment by guardrex** [comment]
@jmprieur ... It's a server API of a hosted Blazor solution. The goal is for the server API to call Graph API after token verification  to get the user's Azure security groups for policies on the API's controllers. In the past, I was **not** using an on-behalf-of approach for the Graph API call. Instead, I was using an application (daemon) approach for the server API. The app was using the user's OID claim and had `Directory.Read.All` permission to make the call. On Friday, I found the daemon docs+examples, so I was going to try that this morning. However ......

> dotnet new blazorwasm2 --auth SingleOrg --calls-graph --hosted

*That looks interesting* ... I'll try that one. Thanks for the tip.

**Comment by guardrex** [comment]
@ChristianWeyer ... I don't think there will be any changes. I haven't been informed of anything. I'll run an upgraded version of the existing IdS hosted test app with 5.0 to confirm.

**Comment by guardrex** [comment]
@jmprieur .... I think that I'm on the right track now with this. My server API acting as a daemon app is now receiving back the user's Azure security groups and roles based on the user's OID claim in their access token and a call to `https://graph.microsoft.com/v1.0/users/{oid}/memberOf`, and that's what I was going for. :tada: Now, I should be able to establish claims for the user on that basis, thus the server API app's auth policies will be effective. Because this approach doesn't use an *on-behalf-of* paradigm for Graph API, it avoids having to grant a client scope for `Directory.Read.All`. Only the server API gets that scope. The user/client app has *least privileged scopes* for the server API ... they only get the registered API scope for the server API.

Let me qualify this approach tho at this *DRAFT* stage ... I'm not a security guru, and security is of great concern to me 😨. We will carefully review this before publishing anything. As I've explored the latest v2.0 docs and sample apps, it looks like user/client scopes for Graph API calls in server API apps are common. It might turn out that an *on-behalf-of* approach would be preferred here. I don't currently favor that approach for this example, but I'll be all :ear: on the PR to see if we need to do something different.

btw - WRT Blazor Server ... My current plan is to leave that to Azure docs and examples given that such apps should align with any ASP.NET Core app, but I need to take a closer look at what I can cross-link (e.g., along the line of what you posted :point_up:).

Thanks again for your help, @jmprieur ... it's always great to have your input! 🎷

**Comment by jmprieur** [comment]
thanks @guardrex 
If you want to ITokenAcqusition directly, you have a method named GetTokenForAppAsync(), which will get you a token for your daemon app

If you want to use the GraphServiceClient, you'd want to read this GitHub issue: https://github.com/AzureAD/microsoft-identity-web/issues/654, which has additional code to make this work for a daemon scenario.

Feel free to send me your code/article to review;

**Comment by guardrex** [comment]
**_I got it!_** :tada:🎈:beers: ..... It just started working all the way through to claims and API access. I am able to use the Graph Service Client and with the acquired token. It seems happy with the setup.

Thanks ... I'll ping you on the PR within a few hours.
